pytorch_blade/pytorch_blade/torch-mlir/lib/Conversion/TorchToMhlo/DiscTorchToMhlo.cpp

```diff
@@ -274,33 +274,158 @@ class ConvertAtenExtractOp : public OpConversionPattern<AtenOpT> {
+      return success();
+    }
+
+    if (elemTy.isa<IntegerType>() && outTy.isa<IntegerType>() && !bitExtend) {
+      auto intMin = rewriter.create<arith::ConstantIntOp>(
+          loc,
+          APInt::getSignedMinValue(outTy.getIntOrFloatBitWidth())
+              .getSExtValue(),
+          elemTy.getIntOrFloatBitWidth());
+
+      auto intMax = rewriter.create<arith::ConstantIntOp>(
+          loc,
+          APInt::getSignedMaxValue(outTy.getIntOrFloatBitWidth())
+              .getSExtValue(),
+          elemTy.getIntOrFloatBitWidth());
+
+      auto smallerThanMin = rewriter.create<arith::CmpIOp>(
+          loc, arith::CmpIPredicate::slt, extractOutput, intMin);
+      auto minOrArg = rewriter.create<arith::SelectOp>(
+          loc, smallerThanMin, intMin, extractOutput);
+      auto largerThanMax = rewriter.create<arith::CmpIOp>(
+          loc, arith::CmpIPredicate::slt, intMax, extractOutput);
+
+      Value clamped = rewriter.create<arith::SelectOp>(
+          loc, largerThanMax, intMax, minOrArg);
+      rewriter.replaceOpWithNewOp<arith::TruncIOp>(op, outTy, clamped);
+      return success();
+    }
+
+    // float to integer
+
+    if (elemTy.isa<FloatType>() && outTy.isInteger(1)) {
+      // Casting to boolean, floats need to only be checked as not-equal to
+      // zero.
+      Value zero = rewriter.create<arith::ConstantOp>(
+          loc, rewriter.getFloatAttr(elemTy, 0.0));
+      rewriter.replaceOpWithNewOp<arith::CmpFOp>(
+          op, arith::CmpFPredicate::UNE, extractOutput, zero);
+      return success();
+    }
+
+    if (arith::FPToSIOp::areCastCompatible(elemTy, outTy)) {
+      auto zero = rewriter.create<arith::ConstantOp>(
+          loc, rewriter.getFloatAttr(elemTy, 0.0f));
+      auto half = rewriter.create<arith::ConstantOp>(
+          loc, rewriter.getFloatAttr(elemTy, 0.5f));
+
+      auto intMin = rewriter.create<arith::ConstantOp>(
+          loc,
+          rewriter.getFloatAttr(
+              elemTy,
+              APInt::getSignedMinValue(outTy.getIntOrFloatBitWidth())
+                  .getSExtValue()));
+
+      auto intMax = rewriter.create<arith::ConstantOp>(
+          loc,
+          rewriter.getFloatAttr(
+              elemTy,
+              APInt::getSignedMaxValue(outTy.getIntOrFloatBitWidth())
+                  .getSExtValue()));
+
+      auto added = rewriter.create<arith::AddFOp>(loc, extractOutput, half);
+      auto subbed = rewriter.create<arith::SubFOp>(loc, extractOutput, half);
+      auto negative = rewriter.create<arith::CmpFOp>(
+          loc, arith::CmpFPredicate::OLT, extractOutput, zero);
+      auto rounded =
+          rewriter.create<arith::SelectOp>(loc, negative, subbed, added);
+
+      Value minValue = rewriter.create<arith::MinFOp>(loc, rounded, intMax);
+      Value clamped = rewriter.create<arith::MaxFOp>(loc, minValue, intMin);
+
+      rewriter.replaceOpWithNewOp<arith::FPToSIOp>(op, outTy, clamped);
+      return success();
```

Comment:
Nit: Add some e2e unit tests for these complicate conversions

Reply:
fptosi type is tested.